### PR TITLE
unit test proof of concept

### PR DESCRIPTION
**Pull request by elharo**
@briandealwis @chanseokoh Anyway to remove/exclude the JUnit dependency? It looks like https://resheim.net/2014/12/unit-testing-eclipse-rcp-applications.html may not have run these from the command line.

### CODE REVIEW

**Comment by briandealwis** [comment]
- should move the `<dependencies>` up — perhaps introduce a /plugins/pom.xml?  Could pull in Mockito too
- can bring in junit and mockito using `build.properties`'s `additional.bundles` or `java.extra.classpath` with `org.junit` and `com.google.cloud.tools.eclipse.test.dependencies`.  The nice thing about these  is that they are preserved when doing an _Plug-in Tools > Update Classpath_ or _Maven > Update Project_.

**Comment by elharo** [comment]
PTAl @briandealwis

**Comment by elharo** [comment]
It works locally but Tycho is having trouble. :-(

**Comment by briandealwis** [comment]
I started down the path of fixing some issues and started hit some package-uses problems.  I'm revising the .test.dependencies and making things better overall.

**Comment by briandealwis** [comment]
I think Tycho 1.0 and Maven 3.5 may have changed how source is processed since Torkild's post, so the approach of adding `source..=src/,src-test/` and `src.excludes=src-test/` no longer works: Maven notices the `src-test` and removes it from the test-source-directory as it's a duplicate, and thus the tests aren't compiled nor executed.  Well actually they are executed sometimes, on the second build if no `clean` was performed.

So we have to only reference the `src-test` from the `.classpath`.  There's a [patch to the m2e-tycho configurators](https://github.com/halwax/m2eclipse-tycho/commit/931183771f5d6555186d4fa7e68219ce028b3fc5) to pick up the `testSourceDirectory` and use it to configure the project classpath.

It turns out we can't put test-dependencies in the top pom, so inherited by all projects, as it leads to package-use conflicts as some junit dependencies are resolved from the Maven-level `junit:junit` and others are from the `com.google.cloud.tools.eclipse.test.dependencies`'s junit.  Putting the test-dependencies directly in the bundles seems to work. 🤞 
🤞

**Comment by briandealwis** [comment]
Oh posted too soon:
```
!ENTRY com.google.cloud.tools.eclipse.test.util 2 0 2017-04-28 20:44:34.560
!MESSAGE Could not resolve module: com.google.cloud.tools.eclipse.test.util [45]
  Unresolved requirement: Import-Package: org.eclipse.swt.widgets
    -> Export-Package: org.eclipse.swt.widgets; bundle-symbolic-name="org.eclipse.swt.win32.win32.x86_64"; bundle-version="3.106.0.v20170126-0932"; version="0.0.0"
       org.eclipse.swt.win32.win32.x86_64 [214]
         Unresolved requirement: Require-Capability: eclipse.platform; filter:="(& (osgi.ws=win32) (osgi.os=win32) (osgi.arch=x86_64))"
         Unresolved requirement: Require-Capability: eclipse.platform; filter:="(& (osgi.ws=win32) (osgi.os=win32) (osgi.arch=x86_64))"
  Bundle was not resolved because of a uses contraint violation.
  org.osgi.service.resolver.ResolutionException: Uses constraint violation. Unable to resolve resource com.google.cloud.tools.eclipse.test.util [osgi.identity; osgi.identity="com.google.cloud.tools.eclipse.test.util"; type="osgi.bundle"; version:Version="0.1.0.201704272116"; singleton:="true"] because it is exposed to package 'org.hamcrest' from resources org.hamcrest.core [osgi.identity; osgi.identity="org.hamcrest.core"; type="osgi.bundle"; version:Version="1.3.0.v201303031735"] and com.google.cloud.tools.eclipse.test.dependencies [osgi.identity; osgi.identity="com.google.cloud.tools.eclipse.test.dependencies"; type="osgi.bundle"; version:Version="0.1.0.201704241845"; singleton:="true"] via two dependency chains.
Chain 1:
  com.google.cloud.tools.eclipse.test.util [osgi.identity; osgi.identity="com.google.cloud.tools.eclipse.test.util"; type="osgi.bundle"; version:Version="0.1.0.201704272116"; singleton:="true"]
    require: (&(osgi.wiring.bundle=org.junit)(bundle-version>=4.12.0))
     |
    provide: osgi.wiring.bundle: org.junit
  org.hamcrest.core [osgi.identity; osgi.identity="org.hamcrest.core"; type="osgi.bundle"; version:Version="1.3.0.v201303031735"]
Chain 2:
  com.google.cloud.tools.eclipse.test.util [osgi.identity; osgi.identity="com.google.cloud.tools.eclipse.test.util"; type="osgi.bundle"; version:Version="0.1.0.201704272116"; singleton:="true"]
    import: (&(osgi.wiring.package=org.mockito)(provider=google)(version>=1.10.19))
     |
    export: osgi.wiring.package: org.mockito; uses:=org.hamcrest
    export: osgi.wiring.package=org.hamcrest
  com.google.cloud.tools.eclipse.test.dependencies [osgi.identity; osgi.identity="com.google.cloud.tools.eclipse.test.dependencies"; type="osgi.bundle"; version:Version="0.1.0.201704241845"; singleton:="true"]
```

I suppose the answer is to remove the `org.hamcrest` from our TPs.

**Comment by elharo** [comment]
So this is working now? I.e. if we put an `Assert.fail()` in a test in src-test then `mvn test` will report the failure and `mvn clean install` errors out?

**Comment by briandealwis** [comment]
I think so.  I'm just on my way out the door, but please feel free to try.

**Comment by briandealwis** [comment]
Just tried and it works for me:
```
[INFO] --- maven-surefire-plugin:2.12.4:test (bundle-unit-test) @ com.google.cloud.tools.eclipse.appengine.standard.java8 ---
[INFO] Surefire report directory: /Users/bsd/Manumitting/Projects/GooglePluginEclipse/repo-gcloud-eclipse/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.google.cloud.tools.eclipse.appengine.standard.java8.AppEngineXsltTransformTest
[…]
Tests run: 9, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.671 sec <<< FAILURE!
testAddBare(com.google.cloud.tools.eclipse.appengine.standard.java8.AppEngineXsltTransformTest)  Time elapsed: 0.005 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[foo]> but was:<[bar]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.google.cloud.tools.eclipse.appengine.standard.java8.AppEngineXsltTransformTest.testAddBare(AppEngineXsltTransformTest.java:47)
[…]
Results :

Failed tests:   testAddBare(com.google.cloud.tools.eclipse.appengine.standard.java8.AppEngineXsltTransformTest): expected:<[foo]> but was:<[bar]>

Tests run: 9, Failures: 1, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[…]
[INFO] com.google.cloud.tools.eclipse.appengine.ui.test ... SUCCESS [  0.325 s]
[INFO] com.google.cloud.tools.eclipse.appengine.standard.java8 FAILURE [  2.964 s]
[…]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:25 min
[INFO] Finished at: 2017-05-01T10:02:20-04:00
[INFO] Final Memory: 193M/414M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.12.4:test (bundle-unit-test) on project com.google.cloud.tools.eclipse.appengine.standard.java8: There are test failures.
```

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src-test/com/google/cloud/tools/eclipse/appengine/deploy/AppEngineDeployOutputTest.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
 public class AppEngineDeployOutputTest {
 
   @Rule public ExpectedException thrown = ExpectedException.none();
-
+
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src-test/com/google/cloud/tools/eclipse/appengine/deploy/AppEngineDeployOutputTest.java]: I can't make this test fail, eithe rin Eclipse or at the command line. It appears to be picking up an older version.

Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src-test/com/google/cloud/tools/eclipse/appengine/deploy/AppEngineDeployOutputTest.java]: Try AppEngineDeployOutputTest. I could not make this test fail, at all.

Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src-test/com/google/cloud/tools/eclipse/appengine/deploy/AppEngineDeployOutputTest.java]: Fails for me: https://travis-ci.org/GoogleCloudPlatform/google-cloud-eclipse/jobs/227602948#L1129

**Comment by briandealwis** [comment]
PTAL @elharo, @chanseokoh.  To summarize:
  - this pulls the various test dependencies like Mockito, Hamcrest, and JUnit, into a single bundle, `com.google.cloud.tools.eclipse.test.dependencies`
  - all test bundles reference this bundle
  - in-bundle unit tests require just a few extra steps to run, which is documented in the `README.md`

The only downside is that it requires modifying the `.classpath`, which is overwritten if you use Maven's _Update Projects_

**Comment by chanseokoh** [comment]
As you said in `README.md` that we can't access Eclipse Core resources (like an Eclipse workspace) and Eclipse UI, I guess in the new experiment in-bundle tests, we can't do UI-related stuff we've been doing a lot, and we can't migrate a large corpus of those existing tests? We will end up having to live with these two test setups , and I wonder if we need some good policy about how we write tests going forward.

**Pull request closed without merging**